### ISSUE TITLE
Probe to understand recommendations for handling a few translated pages

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa#6305]
# Story

As a SiteNow site owner, I would like recommendations on how to currently add translated content to a few pages on my site.

# Additional Info

Admissions currently has some application information translated for Spanish, Chinese, Vietnamese, Swahili, Arabic and French.

We know we currently do not support right to left languages like Arabic.

We can update https://github.com/uiowa/uiowa/issues/6285 after this is complete.

# Questions

- Is there a recommended url structure?
- What is the best practice here so search engines don't see these pages as Duplicate content? We currently allow webmasters to set `<meta http-equiv="content-language"` through our meta tag functionality but the help text says this is Deprecated but still used by bing.
- We have an issue to allow setting the `<html lang` attribute (https://github.com/uiowa/uiowa/issues/5694) but it doesn't sound like that is even what Google wants.
https://developers.google.com/search/docs/specialty/international/localized-versions
- Do we need to do something with sitemap?
- Do we need to do something with `hreflang`?
- Something else? All of the above?

# Proposed solution

**Comment by briand44** [comment]
This is for sites having a small number of translated pages. I would like to avoid the use cases of all content being translated in multiple languages in this slice.

**Comment by bspeare** [comment]
If they want the design of https://admissions.uiowa.edu/arabic to be `rtl` it seems like there are a few options. 

**WYSIWYG only:**
- They could add `<p dir="rtl">Text</p>` in the WYSIWYG and we would need to update our text area styles to work with the `rtl` direction. 

**Block components only:**
- We could add a checkbox on each component to add the `<div dir="rtl">` to each component and we would need to update each component to work with this direction.


**Entire page:**
- The html tag could add the `rtl` direction and then we would be dealing with the header and all components as described in https://github.com/uiowa/uiowa/issues/4649#issuecomment-1017980760.

**Comment by briand44** [comment]
Google info https://developers.google.com/search/blog/2023/09/multilingual-searches?hl=en

**Comment by GaryRidgway** [comment]
# [TLDR HERE](#tldr)
___
# Probe findings
## URL Structure

After reading a few articles [[-1-](https://www.lionbridge.com/blog/global-marketing/choosing-multilingual-url-structure-5-things-consider/)][[-2-](https://www.searchenginejournal.com/subdomains-vs-subfolders-seo/239795/)] about the setup of language based URI structures, it seems that there are a few different ways to structure URIs for different languages (e.g., fr.uiowa.edu, uiowa.fr.edu, uiowa.edu/fr/, uiowa.edu?hl=fr). I will walk through them with small synopsys for  brevity:
1. [Country-Code Top-Level Domain Names](https://www.lionbridge.com/blog/global-marketing/choosing-multilingual-url-structure-5-things-consider/#:~:text=Country%2DCode%20Top%2DLevel%20Domain%20Names%20(ccTLD))  (e.g.  uiowa.fr.edu, or uiowa.fr)
    - This is good for telling a search engine that an entire site is localized for the specific language that is defined. This probably doesn't make sense for us as we are not making entire localized websites.
2. [Subdomains](https://www.lionbridge.com/blog/global-marketing/choosing-multilingual-url-structure-5-things-consider/#:~:text=for%20each%20market.-,Subdomains,-A%20subdomain%20is)  (e.g. fr.uiowa.edu )
    - This Seems like a reasonable option, but it would require different domain structuring, and it also is intended for entirely separate sites that are made entirely localized for a specific language, which is not what we are trying to do.
3. [Subdirectories](https://www.lionbridge.com/blog/global-marketing/choosing-multilingual-url-structure-5-things-consider/#:~:text=country%20or%20language.-,Subdirectories,-The%20most%20commonly)  (e.g. uiowa.edu/french/pageName, or uiowa.edu/fr/pageName)
    - This method means that we don't need to modify certs, and allows us to have different portions of the site that are in different languages. This has the drawback of worse SEO incorporatability, but that seems to be mostly on set up front, and once that is set up is should work pretty smoothly. This also allows us to not have the site be indexed as a only in a specific language, which is VERY important.
4. [URL parameters](https://www.lionbridge.com/blog/global-marketing/choosing-multilingual-url-structure-5-things-consider/#:~:text=A%20fourth%20approach%20is%20to%20use%20query%20strings%20to%20determine%20the%20language%20and%20country.)  (e.g. uiowa.edu/?hl=fr )
    - Sadly I have little linkable evidence for this one, but from what I am gathering google has a really hard time indexing this type of page. It is great for not having to designate your entire site as a language, But seems to be less effective for SEO (harder to set up) and less effective at being indexed.

___

## Google, hreflang
1. `rel="alternate"` paired with the `hreflang` identifier is currently what google seems to be wanting. [[-1-](https://developers.google.com/search/docs/specialty/international/localized-versions)][[-2-](https://www.youtube.com/watch?v=8ce9jv91beQ)]
3. This can be done in three ways  
    1. Through the sitemap, with alternates linked to the base bage
    2. In the HTML `<header>`
    4. In the HTTP header (We will almost assuredly not use this one)
3. I think the sitemap would pretty straightforward to create, however this would probably require some sort of defined taxonomy of languages that are used by the site that could be iterated over to check pages for those languages when constructing the sitemap.
5. The page `<header>` Seems very reasonable, but there would need to either be a link to the alternate page as a field on the translated page, or something of the sort.
    1. References for must be bidirectional. and two separate alternate pages must point to each other with `hreflang`.
6. All pages must list every alternate, including themselves. This goes for both  sitemaps and `<header>`s.
7. [These were tools google offered for debugging](https://developers.google.com/search/docs/specialty/international/localized-versions#debugging-hreflang-errors) `hreflang`

___

## Bing, `http-equiv`
<details><summary>Not a necessary path forward, references kept here for context.</summary>

1. [Bing uses a separate metatag to check things, ](https://toppandigital.com/us/blog-us/what-is-the-hreflang-tag-and-why-is-it-important/#:~:text=It's%20worth%20noting%20that%20while,content%2Dlanguage%E2%80%9D%20meta%20tag.) `http-equiv=“content-language”`
2. [There seems to be some discourse on this.](https://stackoverflow.com/a/4417940)
3. [It also seems to be depreciated?](https://www.w3.org/International/questions/qa-http-and-lang#meta)
4. Okay it seems like we really shouldn't be using this.
5. [Yeah, looks like Bing abandoned that strategy and is using](https://www.bing.com/webmasters/help/webmaster-guidelines-30fba23a#:~:text=If%20you%20have%20multiple%20pages,the%20content%20was%20last%20modified.) `hreflang`.
</details> 

___

## `lang` attribute
1. It seems like, no matter what, we should be wrapping whatever content is actually translated in a container with the appropriate `lang` value. So, for example, if we want to have the header, footer, and menu all be english, but the rest of the content be spanish, we should wrap everything but the aforementioned elements in a container with `lang="es"`.
2. This does mean though that places that are english should also have a `lang="en"` tag.

___

## Dropdown
1. It is recommended that we provide a user with a language dropdown. This could probably be generated by looking for the hreflang tag on the html element and generating a set of pages in other languages for the user to be able to click on. [This was recommended by google in their video on language setting on pages](https://www.youtube.com/watch?v=8ce9jv91beQ&t=976s).

___

## Right to Left text
`dir='rtl'` is [an attribute that can be set to flow text content from left to right in the contents of the tag it is set on](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#dir). It seems a pretty reasonable strategy, and common, to add the `dir='rtl'` attribute to the main content of a page, and leave menus, headers, and footers as english. This seems like it would best be places on the main `<article>` tag. the css selector that most effectively targets this would probably be something like `[dir="rtl"] .region.region-content > .block-system-main-block > article`. The reason this is so specific is because other articles can get inserted in to the page with views and such, and we really want this one. I'm sure there is a better way of doing this, I'm just offering the one I see first.

Once applied, the text will flow in the correct direction, however the styles will not work right now for rtl content. It is probably reasonable to do an audit of styles for rtl, or to allow users to make pages and then report things that they come across to us and to prioritize them as they appear. Either way, we will need an `rtl_overrides.scss` file to contain these overrides for right to left text.

<details><summary>Here are some examples of styles that may need to be tweaked:</summary>
<img src="https://github.com/uiowa/uiowa/assets/23122142/9a8d03ac-d728-4cf3-81be-0e60ff7a8473">
<img src="https://github.com/uiowa/uiowa/assets/23122142/76301eb3-2c1c-4bd4-aad4-81df0d160eb6">
<img src="https://github.com/uiowa/uiowa/assets/23122142/9144a120-11dc-4e10-bc5a-2c4ae09c1d8b">
<img src="https://github.com/uiowa/uiowa/assets/23122142/b4fc8b52-1f1b-4676-a1a4-0e006d3f235f">
</details> 

___

## Language setting
Debatably, the implementation of this is the most non-determined for me. I think we could have a setting on a page that would set the language. We could have a list of languages, their [langcodes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) (e.g. `es` for spanish, `ar` for arabic), and whether they were right to left. Then a hook could fire if it is a language other than english and apply the relevant changes to the page.

Then, we would need a reference field (Only activated and shown if an explicit language was set) that would reference alternatives? But that might get messy if nodes have a lot of alternatives? Its how the `hreflang` stuff works, but maybe there is a better way to do this.

___

# TLDR: Alan's suggestions <a id='tldr'></a>

## Direct answers to questions posed:
___
> Is there a recommended url structure?

We should use subdirectory structure for URL/URIs (e.g. uiowa.edu/LANGCODE/nodeTitle)

___
> What is the best practice here so search engines don't see these pages as Duplicate content?

We should add the alternatives in the `<header>` tag or the sitemap, but not both. Whichever makes the most sense or takes the least effort to implement.

___
## Additional suggestions without direct questions.
___
> Page structure

We should only translate the content of the node, not the header, menus, or footer.


___
> `lang` attribute.

We should have `lang` attributes set around the contents of the page. (e.g. with the Page structure above, we would have `lang='en'` around the header, menus, and footer, and `lang='fr'` around the main content.)


___
> Dropdown

I suggest we have a language dropdown on pages that have multiple languages.


___
> Right to left text.

`dir='rtl'` Should be set in a similar way to the `lang` attribute. We will need new styles for this.


___
>  Language setting

This is the least determined portion of this issue, but I think it could be implemented on a page by page basis.

**Comment by joewhitsitt** [comment]
Thanks for digging into this @GaryRidgway. Just capturing a few details based on the modules we are currently using:

- XML Simple Sitemap suggests that it is already doing what Google wants. _"Sitemaps generated by this module adhere to the new Google standard regarding multilingual content by creating [hreflang sitemaps](https://developers.google.com/search/docs/specialty/international/localized-versions#sitemap) and [image sitemaps](https://developers.google.com/search/docs/crawling-indexing/sitemaps/image-sitemaps) - Googlebots will thank you later."_
- Metatag can be updated to switch from content-language to lang https://www.drupal.org/project/metatag/issues/3217263

**Comment by briand44** [comment]
~Next step, do another probe.~

**Comment by briand44** [comment]
>XML Simple Sitemap suggests that it is already doing what Google wants. 

My understanding is that you have to be using the core language modules for XML Simple Sitemap to behave the way Google wants.

>The sitemap is multilingual out of the box. As a hreflang sitemap, it provides links to all language variants from only one URL. See [this](https://gbyte.dev/sitemap.xml) example. You need to have multiple languages enabled on the Drupal instance for this to work.

>Metatag can be updated to switch from content-language to lang https://www.drupal.org/project/metatag/issues/3217263

Isn't this just removing content-language?

i did a little digging into this today and not seeing many easy wins here without adopting the core multilingual modules.

**Comment by briand44** [comment]
There does seem to be a metatag submodule for hreflang though.
